src/Applications.hh

```diff
@@ -1,91 +1,527 @@
-
 #ifndef APPLICATIONS_DEF
 #define APPLICATIONS_DEF
 
 #include <map>
+#include <unordered_map>
+#include <functional>
+#include <string>
+#include <set>
+#include <list>
+#include <type_traits>
+#include "Formatters.hh"
+#include "Application.hh"
+
+// a container friendly version of const std::string &
+struct string_ref : public std::reference_wrapper<const std::string>
+{
+    using std::reference_wrapper<const std::string>::reference_wrapper;
+
+    bool operator<(const string_ref & other) const
+    {
+        return get() < other.get();
+    }
+
+    bool operator==(const string_ref & other) const
+    {
+        return get() == other.get();
+    }
+};
+
+template <>
+struct std::hash<string_ref>
+{
+    std::size_t operator()(const string_ref & tohash) const noexcept
+    {
+        return std::hash<remove_const<string_ref::type>::type>{}(tohash.get());
+    }
+};
 
 #include "Application.hh"
 
-class Applications : public std::map<std::string, Application*>
+// a helper function that safely gets an iterator from container.insert() and container.emplace()
+template <typename iterator>
+iterator getiter(const std::pair<iterator, bool> & p)
+{
+    if (!p.second) {
+        fprintf(stderr, "Tried to insert an element when it was already present, aborting...\n");
+        abort();
```

Comment:
Why not throw an exception instead?

Reply:
I am unsure because I haven't read the code thoroughly yet but it looks like the condition is fatal enough that there is no recovery possible and catching the exception wouldn't make sense. I choose `abort()` in code that I write if some internal invariant is broken.